File: v8-core/src/experts/fib_retracement_continuation.rs
//! fib_retracement_continuation: evaluate() port target (issue #85) — mirror src/v8/experts/fib_retracement_continuation.py
//! bit-for-bit (PARITY_AND_IDENTITY_SPEC §3; COMPUTE_CORE_SPEC §8 S4).
//!
//! Variant 'a' (ratio 0.382): a pullback that reaches the retracement level
//! (low <= level) and is reclaimed by the close (close > level) on an up-
//! impulse, mirrored for down-impulses. The deepest retracement (0.786) is the
//! invalidation reference, frozen into the draft geometry as prior_low_ref /
//! prior_high_ref (D-042, `prior_*_ref` pattern). Anchor is the standard
//! D-026 base find_setup_anchor over the pullback-reclaim predicate.

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::HistBar;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["location", "volatility", "history"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const TARGET_R: f64 = 1.0;
pub const _STOP_R: f64 = 1.0;
pub const EXPIRY_BARS: i64 = 8;

/// The frozen variant 'a' retracement ratio (self._RATIO['a']).
const _RATIO: f64 = 0.382;
/// The deepest retracement: the invalidation / stop reference level.
const DEEP_RETRACEMENT: f64 = 0.786;

/// Python `_retracement_level(fibs, ratio)`: the level for `ratio` from the
/// self-describing fib array [extreme, direction, retr, ext]; None when the
/// shape is wrong or no pair matches within 1e-9.
fn retracement_level(fibs: &serde_json::Value, ratio: f64) -> Option<f64> {
    let arr = fibs.as_array()?;
    if arr.len() != 4 {
        return None;
    }
    for pair in arr[2].as_array()? {
        let p = match pair.as_array() {
            Some(p) if p.len() == 2 => p,
            _ => return None,
        };
        let r = p[0].as_f64()?;
        if (r - ratio).abs() < 1e-9 {
            return p[1].as_f64();
        }
    }
    None
}

/// Python `_long_pred`: bar i reached the retracement level and reclaimed it
/// by the close (pullback-reclaim of an up-impulse).
fn long_pred(i: usize, b: &HistBar, level: f64) -> bool {
    i > 0 && b.close > level && b.low <= level
}

/// Python `_short_pred`: mirror (price rose to the level, closed back below).
fn short_pred(i: usize, b: &HistBar, level: f64) -> bool {
    i > 0 && b.close < level && b.high >= level
}

pub fn fib_retracement_continuation(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    let sym = fm.symbol;
    // `_need` over {sym}.close/.atr/.history/.fib_levels; then the None guards.
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let atr = match fm.value("atr") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let fibs = match fm.features.get("fib_levels") {
        Some(f) => f.value.clone(),
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    // hist must be a non-empty (tuple|list) — empty window is NO_HABITAT.
    if fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let fibs_arr = match fibs.as_array() {
        Some(a) if a.len() == 4 => a,
        _ => return no_habitat(expert_id, version, fm.as_of),
    };
    let anchor_price = match fibs_arr[0].as_f64() {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let direction = match fibs_arr[1].as_f64() {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    // The fib anchor's own consistency is the guard (degenerate pair => None).
    if !(direction == 1.0 || direction == -1.0) || anchor_price <= 0.0 {
        return no_habitat(expert_id, version, fm.as_of);
    }

    if version == "v1" {
        let level = match retracement_level(&fibs, 0.382) {
            Some(v) => v,
            None => return no_habitat(expert_id, version, fm.as_of),
        };
        let deep = match retracement_level(&fibs, DEEP_RETRACEMENT) {
            Some(v) => v,
            None => return no_habitat(expert_id, version, fm.as_of),
        };
        let (direction_sig, pred): (&str, fn(usize, &HistBar, f64) -> bool) = if direction == 1.0 {
            ("LONG", long_pred)
        } else {
            ("SHORT", short_pred)
        };
        let n = fm.history.len();
        if !pred(n - 1, &fm.history[n - 1], level) {
            return no_setup(expert_id, version, fm.as_of);
        }
        let anchor = find_setup_anchor(&fm.history, &|i, b| pred(i, b, level));
        let draft = Draft {
            direction: direction_sig.to_string(),
            birth_time: fm.as_of,
            risk_geometry: geom(vec![
                ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
                ("target_r", serde_json::json!(TARGET_R)),
                ("stop_r", serde_json::json!(1.0)),
                ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
                ("atr_ref", serde_json::json!(atr)),
                (
                    if direction_sig == "LONG" {
                        "prior_low_ref"
                    } else {
                        "prior_high_ref"
                    },
                    serde_json::json!(deep),
                ),
            ]),
        };
        let fingerprint = format!("{sym}:{:.3}:{:.6}:{direction_sig}", 0.382, level);
        return candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint);
    }

    let ratios = [0.382, 0.500, 0.618];
    let mut hit_level: Option<(f64, f64)> = None; // (ratio, level_price)

    let n = fm.history.len();
    let newest = &fm.history[n - 1];

    for &r in &ratios {
        if let Some(lvl) = retracement_level(&fibs, r) {
            let reached_and_reclaimed = if direction == 1.0 {
                newest.low <= lvl && newest.close > lvl
            } else {
                newest.high >= lvl && newest.close < lvl
            };
            if reached_and_reclaimed {
                hit_level = Some((r, lvl));
                break;
            }
        }
    }

    let (active_ratio, level) = match hit_level {
        Some(hl) => hl,
        None => return no_setup(expert_id, version, fm.as_of),
    };

    let deep = match retracement_level(&fibs, DEEP_RETRACEMENT) {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let (direction_sig, pred): (&str, fn(usize, &HistBar, f64) -> bool) = if direction == 1.0 {
        ("LONG", long_pred)
    } else {
        ("SHORT", short_pred)
    };

    let anchor = find_setup_anchor(&fm.history, &|i, b| pred(i, b, level));
    let raw_stop_r = (close - deep).abs() / atr;
    let stop_r = raw_stop_r.clamp(0.8, 2.0);

    let draft = Draft {
        direction: direction_sig.to_string(),
        birth_time: fm.as_of,
        risk_geometry: geom(vec![
            ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
            ("target_r", serde_json::json!(TARGET_R)),
            ("stop_r", serde_json::json!(stop_r)),
            ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
            ("atr_ref", serde_json::json!(atr)),
            // The invalidation reference is FROZEN at detection (prior_*_ref,
            // D-042): prior_low_ref for LONG, prior_high_ref for SHORT.
            (
                if direction_sig == "LONG" {
                    "prior_low_ref"
                } else {
                    "prior_high_ref"
                },
                serde_json::json!(deep),
            ),
            ("ratio", serde_json::json!(active_ratio)),
        ]),
    };
    let fingerprint = format!("{sym}:{:.3}:{:.6}:{direction_sig}", active_ratio, level);
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}

File: src/v8/experts/fib_retracement_continuation.py
"""Fibonacci retracement continuation behavior family (`fib_retracement_continuation`).

Hypothesis (Ch10.4-10.7, Ch10.14): inside an intact impulse, a pullback that
reaches a Fibonacci retracement level of the prior impulse range and is then
reclaimed by a close is a continuation setup. Detected on closed bars only;
the direction follows the impulse (LONG after an up-impulse retracement
reclaim, SHORT after a down-impulse retracement reclaim).

The retracement level set is `{sym}.fib_levels`, a self-describing tuple
(anchor_price, direction, retracements, extensions) anchored on the most
recent CONFIRMED swing pair (G-24). The area-of-application invariant
(Ch10.11 p394: levels are valid only after the anchor bar) holds by
construction: the anchor is always a past confirmed swing, so the levels are
only ever applied forward of it.

The deepest retracement (78.6%) is the invalidation reference: a close beyond
it is a deep correction — "the deeper the correction, the lower the
probability of trend resumption" (Ch10.14) — and the level set no longer
applies. It is frozen into the draft geometry as prior_low_ref / prior_high_ref
(the `prior_*_ref` pattern, D-042), excluded from the episode-key geometry.

Risk geometry deviation (documented): the book's stop ("just below the 78.6%
level") and target ("1:1 projection of the impulse range") are PRICE LEVELS.
V8's canonical simulator expresses stops/targets as fixed R-multiples of a
declared risk unit (D-028), and a data-derived stop_r would break episode-key
stability (D-026) by hashing differently across decision clocks. The levels
are therefore frozen as the prior_*_ref used by still_valid and the pre-entry
invalidation, while the structural geometry is the family default 1R:1R:8bar
with atr_ref — exactly how the registered pilots carry their book level rules
(failed_breakout's prior_high_ref).

Rule-13 discipline: retracement-ratio and direction changes are VARIANTS of
this one behavior family, never separate Experts; variants_evaluated lists
every ratio variant implemented and tested (D-044).
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert

# Retracement ratios of the prior impulse range (G-24, verbatim from the book).
FIB_RETRACEMENTS = (0.236, 0.382, 0.5, 0.618, 0.786)
# The deepest retracement: the invalidation / stop reference level.
DEEP_RETRACEMENT = 0.786


class FibRetracementContinuationExpert(Expert):
    """Pullback to a retracement level reclaimed by a close, in the direction
    of the intact impulse."""
    expert_id = 'fib_retracement_continuation'
    version = 'v1'
    mechanism_family_id = 'fib_level_reaction'
    behavior_family_id = 'fib_retracement_continuation'
    variant_id = 'a'
    requires = ('location', 'volatility', 'history')
    # D-044: every retracement-ratio variant implemented and tested in
    # tests/test_expert_fib_retracement_continuation.py.
    variants_evaluated = ['a', 'b', 'c', 'd', 'e']
    # D-046: the search consumed 5 retracement ratios x 2 direction-sign
    # choices. Every threshold was frozen in code against crafted tapes before
    # any real window existed; nothing was tuned on data.
    search_universe_size = 10
    _RATIO = {'a': 0.382, 'b': 0.5, 'c': 0.618, 'd': 0.236, 'e': 0.786}

    @staticmethod
    def _retracement_level(fibs: tuple, ratio: float):
        """The retracement level for `ratio` from the self-describing fib
        tuple (anchor, direction, retr, ext); None when absent."""
        if not isinstance(fibs, tuple) or len(fibs) != 4:
            return None
        for r, level in fibs[2]:
            if abs(r - ratio) < 1e-9:
                return float(level)
        return None

    def _long_pred(self, i: int, bar: tuple) -> bool:
        """Bar i reached the retracement level (low <= level) and reclaimed it
        by the close (close > level) — the pullback-reclaim of an up-impulse."""
        if i == 0:
            return False
        _e, _o, _h, low, close, _f, _s = bar
        return close > self._level and low <= self._level

    def _short_pred(self, i: int, bar: tuple) -> bool:
        """Mirror: price rose to the retracement level and closed back below it."""
        if i == 0:
            return False
        _e, _o, high, _l, close, _f, _s = bar
        return close < self._level and high >= self._level

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.history',
                f'{sym}.fib_levels']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        fibs = f[f'{sym}.fib_levels'].value
        hist = f[f'{sym}.history'].value
        if atr is None or not isinstance(fibs, tuple) or len(fibs) != 4 \
                or not isinstance(hist, (tuple, list)) or not hist:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        # The fib anchor's own consistency is the guard: `_fib_levels` returns
        # None when the confirmed swing pair is degenerate, and the level set
        # is checked below. The significance-filtered swing_high_10/
        # swing_low_10 features are a DIFFERENT pair (CRIT-1 range filter) than
        # the unfiltered confirmed swings the anchor uses — gating on them
        # vetoed states with a valid anchor and NO_HABITAT'd states where the
        # filtered pair was absent entirely.
        anchor_price, direction, _retr, _ext = fibs
        if direction not in (1.0, -1.0) or anchor_price <= 0.0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        level = self._retracement_level(fibs, self._RATIO[self.variant_id])
        deep = self._retracement_level(fibs, DEEP_RETRACEMENT)
        if level is None or deep is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        # LONG on an up-impulse (direction +1), SHORT on a down-impulse (-1).
        # "Trend in effect" is the intact IMPULSE itself: the anchor direction
        # is the trend, and the 78.6% depth gate below (frozen as the
        # prior_*_ref) is the "impulse still intact" check — a short/long EMA
        # cross would fight the family's own structure (a deep pullback
        # naturally puts EMA5 below EMA20 while the impulse is fine).
        if direction == 1.0:
            direction_sig = 'LONG'
            pred = self._long_pred
        else:
            direction_sig = 'SHORT'
            pred = self._short_pred
        self._level = level
        hist = tuple(hist)
        if not pred(len(hist) - 1, hist[-1]):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        anchor = self.find_setup_anchor(hist, pred)
        risk_geometry = self.declared_geometry()
        risk_geometry['atr_ref'] = atr
        # The invalidation reference is FROZEN at detection (the `prior_*_ref`
        # pattern, D-042): a live-recomputed deep level would drift with the
        # adverse move and the dead-thesis close would never fire.
        if direction_sig == 'LONG':
            risk_geometry['prior_low_ref'] = deep
        else:
            risk_geometry['prior_high_ref'] = deep
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction_sig,
            setup_fingerprint=f'{sym}:{self._RATIO[self.variant_id]:.3f}:'
                              f'{level:.6f}:{direction_sig}',
            risk_geometry=risk_geometry,
            birth_time=t, setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The thesis is "the pullback reclaimed its retracement level and the
        impulse continues". A close beyond the deepest retracement (78.6%) is a
        deep correction — the impulse is no longer intact, so the reason to hold
        is gone whatever the stop distance still says (Ch10.14; E-12
        INVALIDATION). Uses the FROZEN detection-time level, not the live one.
        Unobservable inputs fail open (an unreadable thesis is not a dead one)."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True
        if draft.direction == 'LONG':
            ref = draft.risk_geometry.get('prior_low_ref')
            if ref is None:
                return True
            return float(close.value) > float(ref)
        ref = draft.risk_geometry.get('prior_high_ref')
        if ref is None:
            return True
        return float(close.value) < float(ref)
